Faulty IRC reconnect logic (for the billionth time)

(2024-03-18)

> [7:55 AM] Laurakko: this supposed to happen when bancho restarts? the reconnecting timer just goes on forever
> ![image](https://github.com/TicClick/steel/assets/4686101/c4cea0c9-9881-4656-94ba-d3a6e650a737)
> 
> [8:40 AM] TicClick: not at all — can you send me the logs?
> [8:41 AM] Laurakko: just runtime.log?
> [8:42 AM] Laurakko:
> ```
> 2024-03-10 21:53:36.9495358 [WARN] failed to remove old executable ("D:\\osu!\\nonosuthings\\steel irc\\steel.exe.old") which was left after SUCCESSFUL update: Os { code: 5, kind: PermissionDenied, message: "Access is denied." }
> 2024-03-13 06:52:53.6917957 [ERROR] IRC chat error: FatalError("connection broken: an io error occurred")
> 2024-03-15 09:20:25.3787392 [ERROR] IRC chat error: FatalError("connection broken: connection reset: no ping response")
> 2024-03-16 10:27:01.0520054 [ERROR] IRC chat error: FatalError("connection broken: connection reset: no ping response")
> 2024-03-16 22:15:36.4499511 [ERROR] IRC chat error: FatalError("connection broken: connection reset: no ping response")
> ```
> [8:42 AM] Laurakko: idk there doesn't seem to be anything for today
> [8:42 AM] TicClick: yeah, will have to find this out on my own
> [8:43 AM] Laurakko: tbf this is probably rare, since bancho doesn't fully restart a lot
> ![image](https://github.com/TicClick/steel/assets/4686101/e8684d89-e6b8-475b-99db-f84d495f851a)
> 
> [8:44 AM] Laurakko: it's still going, maybe it would be beneficial to add a button to attempt reconnecting?
> ![image](https://github.com/TicClick/steel/assets/4686101/7e37bbe9-d058-41e6-b781-71cad425b5c2)
> 
> [9:14 AM] TicClick: sure is — please restart it manually for now
> [9:14 AM] Laurakko: yea i did
> [9:14 AM] TicClick: it's worse, since it's general reconnect logic, not bancho restart-specific
> [9:14 AM] Laurakko: ah

(2024-04-05)

> [10:12 PM] Laurakko: pain
> ![image](https://github.com/TicClick/steel/assets/4686101/edc47468-56a9-4bff-af9e-266a11664c67)
> [10:12 PM] Laurakko: again nothing in the log
> [10:19 PM] TicClick: ehh, gotta work on it in the next few days
> [10:19 PM] Laurakko: but yea, only happens when bancho restarts

--- a/src/core/irc/actor.rs
+++ b/src/core/irc/actor.rs
@@ -120,6 +120,7 @@ impl IRCActor {
                         .unwrap();
                     *irc_stream_sender.lock().unwrap() = Some(clt.sender());
 
+                    let mut disconnected_by_user = false;
                     let mut stream = clt.stream().unwrap();
                     loop {
                         let (lock, cv) = &*arc;
@@ -130,10 +131,7 @@ impl IRCActor {
                             if *g {
                                 *g = false;
                                 clt.send_quit("").unwrap();
-                                tx.blocking_send(AppMessageIn::ConnectionChanged(
-                                    ConnectionStatus::Disconnected { by_user: true },
-                                ))
-                                .unwrap();
+                                disconnected_by_user = true;
                                 break;
                             }
                         }
@@ -148,10 +146,6 @@ impl IRCActor {
                                         )),
                                     ))
                                     .unwrap();
-                                    tx.blocking_send(AppMessageIn::ConnectionChanged(
-                                        ConnectionStatus::Disconnected { by_user: false },
-                                    ))
-                                    .unwrap();
                                     break;
                                 }
                                 Ok(msg) => {
@@ -159,14 +153,21 @@ impl IRCActor {
                                 }
                             },
                             None => {
-                                tx.blocking_send(AppMessageIn::ConnectionChanged(
-                                    ConnectionStatus::Disconnected { by_user: false },
-                                ))
+                                tx.blocking_send(AppMessageIn::ChatError(IRCError::FatalError(
+                                    "remote server has closed the connection, probably because it went offline".to_owned(),
+                                )))
                                 .unwrap();
                                 break;
                             }
                         }
                     }
+
+                    tx.blocking_send(AppMessageIn::ConnectionChanged(
+                        ConnectionStatus::Disconnected {
+                            by_user: disconnected_by_user,
+                        },
+                    ))
+                    .unwrap();
                 }
             }
         }));

--- a/src/gui/menu.rs
+++ b/src/gui/menu.rs
@@ -122,7 +122,7 @@ impl Menu {
                 ConnectionStatus::InProgress => ("connecting...".to_owned(), false),
                 ConnectionStatus::Scheduled(when) => {
                     let action = format!(
-                        "reconnecting ({}s)",
+                        "reconnect (or wait {}s)",
                         (when - chrono::Local::now()).num_seconds()
                     );
                     (action, true)

